### PR TITLE
Implement server-side messages

### PR DESCRIPTION
**Pull request by nabijaczleweli**
This change is [Reviewable](https://reviewable.io/reviews/loungecpp/tatsoryk/14)

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
Still not sure what to do with `world_state` without official `Player` and `Bullet` types

**Comment by TheCatPlusPlus** [comment]
Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

_[server/src/message.rs, line 302 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-vgkG2VKRWj_KJ9rW:-KE-vgkHCPjFSKyi5Iso:53067045) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L302)):_
Avoid unwraps. This doesn't have to be inside `to_string`.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE-wk-65_SQe_kMEylT:-1629932950) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
Serde seems to support `#[derive]`, not sure if all the manual fiddling is necessary.

---

_[server/src/message.rs, line 710 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-xDknpxHUI4xllBRv:-KE-xDknpxHUI4xllBRw:-1571266309) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L710)):_
This `&*&` is really silly and unnecessary :v `&json_txt` should be enough, `&json_txt[..]` if compiler needs help with inference.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by TheCatPlusPlus** [comment]
> Still not sure what to do with world_state without official Player and Bullet types

They're defined in the spec.

**Comment by nabijaczleweli** [comment]
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

_[server/src/message.rs, line 302 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-vgkG2VKRWj_KJ9rW:-KE0ep4zHOD3rj59P5BF:-1427051060) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L302)):_
`to_string`ing `serde_json::Value` cannot fail, though.
As for being inside `to_string`, see L308 discussion.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE0eNctcMG3oYAWQmhU:-1446774293) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
I was thinking about abstracting `serde[_json]` out of this entirely, and Rust stdlib provides all the right tools.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by nabijaczleweli** [comment]
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

_[server/src/message.rs, line 710 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-xDknpxHUI4xllBRv:-KE0gDcgQmZVpc6cKcdx:-499146009) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L710)):_
Done.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by TheCatPlusPlus** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

_[server/src/message.rs, line 302 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-vgkG2VKRWj_KJ9rW:-KE0hTyhDgblZ80np_6r:2054165560) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L302)):_
Well if it's guaranteed to not fail then fine. Maybe add a note.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE0gadl5DzKomEol5dj:-813743864) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
Serde already is an abstraction, I don't think it's worth to be able to switch serialisation frameworks (as long as this one is reasonably well-written; if not, then let's switch immediately). It's better to use Serialize/Deserialize and their annotations to control behaviour instead of ToString/FromStr.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by nabijaczleweli** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE0lWafneUBN4qaFcMP:168684241) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
Nota Bene: if we're supporting Rust Stable, then we don't have [`custom_derive`](https://github.com/rust-lang/rust/issues/29644), which means we don't benefit from automatic de-/serialization macros.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by TheCatPlusPlus** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE0mXibKvVTJIFyBJoa:294745738) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
https://github.com/serde-rs/serde#using-serde-with-stable-rust-syntex-and-serde_codegen (`custom_derive` is only syntax sugar anyway, the real necessary feature is compiler plugins)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by TheCatPlusPlus** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

_[server/src/message.rs, line 308 [r1]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE-wk-59r28hIYHWoUJ:-KE0wAY0z_fjPJ1Po-_j:317389731) ([raw file](https://github.com/loungecpp/tatsoryk/blob/a76d959960570264c1e6ee2dfb138f4e6e5e298c/server/src/message.rs#L308)):_
We're leaving that for now.

---

_[server/src/message.rs, line 203 [r2]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE0w5ZumCTzrsvAU0sq:-KE0w5ZumCTzrsvAU0sr:-1547184712) ([raw file](https://github.com/loungecpp/tatsoryk/blob/dd6abb1168d3a92ccda1cd42ebddee10f2e08966/server/src/message.rs#L203)):_
Marking todo.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by TheCatPlusPlus** [comment]
Reviewed 3 of 4 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

_[server/src/message/mod.rs, line 210 [r3]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE0yyi5KjKbSxHHd2zL:-KE0yyi5KjKbSxHHd2zM:-1547184712) ([raw file](https://github.com/loungecpp/tatsoryk/blob/5d40ed6d44b8c1f9039253335e776600e921c49d/server/src/message/mod.rs#L210)):_
Marking todo.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by thecoshman** [comment]
Are we going to go for squashing commits down before PR is finished?

**Comment by TheCatPlusPlus** [comment]
Might fuck up review status so we'll see after it's all reviewed.

File: server/src/message/mod.rs [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+    ///   - `move_x` (Optional&lt;f32&gt;) — current movement vector X of the player, if player is moving
+    ///   - `move_y` (Optional&lt;f32&gt;) — current movement vector Y of the player, if player is moving
+    /// - `alive_bullets` (Bullet[]) — an array of all currently alive bullets, each containing:
+    ///   - `id` (u32) — ID of the bullet
+    ///   - `x` (f32) — current position X of the bullet
+    ///   - `y` (f32) — current position Y of the bullet
+    ///   - `move_x` (f32) — current movement vector X of the bullet
+    ///   - `move_y` (f32) — current movement direction vector Y of the bullet
+    ///                      (movement direction vectors MUST be normalised, i.e. their magnitude MUST be equal to 1)
+    WorldState,
```
Comment (thecoshman) [inline server/src/message/mod.rs]: What's this here?

File: server/src/message/mod.rs [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+    Fire {
+        move_x: f32,
+        move_y: f32,
+    },
+}
+
+impl ToString for Message {
+    fn to_string(&self) -> String {
+        let mut values = BTreeMap::new();
+        let msg_type = match self {
```
Comment (thecoshman) [inline server/src/message/mod.rs:246]: Is this really the best structure? Having to come here for every change to messages is going to. Can't each individual message cleanly implement it's own 'to_string'?

File: server/src/message/mod.rs [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+                                serde_json::Value::String(msg_type.to_string()));
+        if !values.is_empty() {
+            let _ = root_obj.insert("data".to_string(), serde_json::Value::Object(values));
+        }
+
+        serde_json::to_string(&serde_json::Value::Object(root_obj)).unwrap()
+    }
+}
+
+impl FromStr for Message {
```
Comment (thecoshman) [inline server/src/message/mod.rs:328]: Again, urgh, such monstrously long methods. Every type has to be handled it's own way... why cram it all together like this?

File: server/src/message/mod.rs (hunk 2 above)
Comment (nabijaczleweli) [inline server/src/message/mod.rs:246]: It is cleanly implemented.

File: server/src/message/mod.rs (hunk 1 above)
Comment (nabijaczleweli) [inline server/src/message/mod.rs]: TODO

File: server/src/message/mod.rs (hunk 3 above)
Comment (nabijaczleweli) [inline server/src/message/mod.rs:328]: There's one type here, `Message`

File: server/src/message/mod.rs (hunk 2 above)
Comment (thecoshman) [inline server/src/message/mod.rs:246]: It's not though, Open-Closed :\

File: server/src/message/mod.rs (hunk 3 above)
Comment (thecoshman) [inline server/src/message/mod.rs:328]: All of the types of message though... or are they not each an individual type? If not, why not?

Comment (nabijaczleweli) [inline server/src/message/mod.rs:328]: Why would they be? How would we deserialise them then?

File: server/src/message/mod.rs (hunk 2 above)
Comment (nabijaczleweli) [inline server/src/message/mod.rs:246]: > Open-Closed 

вот

Comment (TheCatPlusPlus) [inline server/src/message/mod.rs:246]: Open/closed principle is garbage.

File: server/src/message/mod.rs (hunk 3 above)
Comment (TheCatPlusPlus) [inline server/src/message/mod.rs:328]: Single enum is better. Processing code will be centralised anyway.

Comment (andyprowl) [inline server/src/message/mod.rs:328]: I'd keep serialization logic separate from message type(s). SRP and stuff.

Comment (TheCatPlusPlus) [inline server/src/message/mod.rs:328]: The whole purpose of this enum is serialisation. Plus it's implementation of traits, it's as separate as it can be.

Comment (andyprowl) [inline server/src/message/mod.rs:328]: I wasn't disagreeing. I was addressing Pirate's comment.

File: server/src/message/mod.rs (hunk 1 above)
Comment (thecoshman) [inline server/src/message/mod.rs]: slacker

Comment (nabijaczleweli) [inline server/src/message/mod.rs]: I'm doing it

**Comment by nabijaczleweli** [comment]
Review status: 4 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

_[server/src/message/mod.rs, line 210 [r3]](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-KE0yyi5KjKbSxHHd2zL:-KE2Na4sgtXXaZePtjTK:-499146009) ([raw file](https://github.com/loungecpp/tatsoryk/blob/5d40ed6d44b8c1f9039253335e776600e921c49d/server/src/message/mod.rs#L210)):_
Done.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14)_

**Comment by nabijaczleweli** [comment]
Not sure why Travis is failing, looks like `clang-3.8` package magically stopped containing Clang binaries?

**Comment by TheCatPlusPlus** [comment]
Seriously don't rebase shit before review is over, or it's hard to tell what was already read and what wasn't. (And impossible to tell what actually changed from iteration to iteration)

Okay I can actually get normal diffs, but don't anyway, it can be done afterwards.

**Comment by TheCatPlusPlus** [comment]
:+1:

---

Reviewed 1 of 3 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/loungecpp/tatsoryk/14#-:-KE4tx69pKP6PZTal-Jl:676269983)_

**Comment by thecoshman** [comment]
:+1:

**Comment by nabijaczleweli** [comment]
2 upboats, merging

**Pull request merged**